### ISSUE TITLE
crdb conditional upsert (`insert_into ... on_conflict(..).filter_target(...).do_update()...`) doesn't apply the `filter_target` predicate

### ISSUE THREAD

**Issue by sunshowers** [oxidecomputer/omicron#5047]
tl;dr: let's say you want to:

* upsert a row into a database
* but only update it if a particular condition matches

The suggested way to do this is with Diesel's [`DecoratableTarget::filter_target`](https://docs.diesel.rs/master/diesel/prelude/trait.DecoratableTarget.html#tymethod.filter_target), but if you try and use that, CockroachDB silently drops the condition associated with `filter_target`. This is pretty bad.

---

### More details

I discovered this while working on #5032.

The [Postgres documentation on INSERT](https://www.postgresql.org/docs/current/sql-insert.html) documents two possible formulations for a conditional upsert:

1. `ON CONFLICT ... WHERE ... DO UPDATE`
2. `ON CONFLICT ... DO UPDATE ... WHERE`

`filter_target` becomes 1, not 2. There appears to be no way in stock Diesel to write 2, though it should be possible to do so with an extension.

CockroachDB silently ignores the `WHERE` clause in queries like 1, though it correctly recognizes 2.

[See this gist for a repro](https://gist.github.com/sunshowers/92a7abb61452a30d2520d900589737f3#file-gistfile1-txt-L25).

As the comment there indicates, you can even pass in a completely non-existent column -- and Cockroach completely ignores that.

---

To do:

- [ ] Figure out whether the bug still exists in the latest version of Cockroach
- [ ] If so, file a bug upstream
- [ ] Audit current uses of `DecoratableTarget::filter_target` since they don't work
- [ ] Write a Diesel extension that implements the `ON CONFLICT ... DO UPDATE ... WHERE` pattern
- [ ] Switch over current uses of `DecoratableTarget::filter_target` to that.
- [ ] Add clippy lint to ensure that `DecoratableTarget::filter_target` isn't used.
